cpp/turbodbc_arrow/Library/src/set_arrow_parameters.cpp

```diff
@@ -0,0 +1,375 @@
+#include <turbodbc_arrow/set_arrow_parameters.h>
+
+#include <arrow/python/pyarrow.h>
+
+#include <turbodbc/errors.h>
+#include <turbodbc/make_description.h>
+#include <turbodbc/time_helpers.h>
+
+#ifdef _WIN32
+#include <windows.h>
+#endif
+#include <sql.h>
+
+#include <ciso646>
+
+using arrow::BooleanArray;
+using arrow::BinaryArray;
+using arrow::ChunkedArray;
+using arrow::Date32Array;
+using arrow::DoubleType;
+using arrow::Int64Type;
+using arrow::NumericArray;
+using arrow::StringArray;
+using arrow::Table;
+using arrow::TimestampArray;
+using arrow::TimestampType;
+using arrow::TimeUnit;
+using arrow::TypeTraits;
+
+namespace turbodbc_arrow {
+
+namespace {
+
+    struct parameter_converter {
+        parameter_converter(std::shared_ptr<ChunkedArray> const & data,
+                            turbodbc::bound_parameter_set & parameters,
+                            std::size_t parameter_index) :
+            data(data),
+            parameters(parameters),
+            parameter_index(parameter_index)
+        {}
+
+        cpp_odbc::multi_value_buffer & get_buffer() {
+            return parameters.get_parameters()[parameter_index]->get_buffer();
+        }
+
+        template <size_t element_size>
+        void set_indicator(cpp_odbc::multi_value_buffer & buffer, int64_t start, int64_t elements) {
+            if (data->num_chunks() != 1) {
```

Comment:
This check is made many times in the code. Please consider making the check elsewhere (in calling code, for example) or put the check in the parameter_converter class once and use the NVI pattern to enforce the check for subclasses

Reply:
The highest point would probably be the Python layer. That is also the place where checks for the related `c_contiguous` flag is done for NumPy arrays: https://github.com/blue-yonder/turbodbc/blob/master/python/turbodbc/cursor.py#L55

The Python layer also has the advantage of being a single code path (instead of six) that is easily testable with a single test ;-).